kybra/compiler/kybra_generate/src/guard_function/mod.rs

```diff
@@ -23,25 +46,30 @@ impl SourceMapped<&Located<StmtKind>> {
         }
     }
 
-    pub fn as_guard_function(&self) -> KybraResult<Option<GuardFunction>> {
+    pub fn as_guard_function(&self) -> Result<Option<GuardFunction>, Vec<Error>> {
         if !self.is_guard_function() {
             return Ok(None);
         }
         match &self.node {
             StmtKind::FunctionDef { name, .. } => {
-                if self.has_params() {
-                    return Err(self.guard_functions_param_error());
-                }
-                if !self.returns_guard_result() {
-                    return Err(self.guard_function_return_error());
-                }
+                (
+                    match !self.has_params() {
+                        true => Ok(()),
+                        false => Err(GuardFunctionParam::err_from_stmt(self).into()),
+                    },
+                    match self.returns_guard_result() {
+                        true => Ok(()),
+                        false => Err(GuardFunctionReturn::err_from_stmt(self).into()),
+                    },
+                )
+                    .collect_results()?;
 
                 Ok(Some(GuardFunction {
                     body: guard_function::generate(name),
                     name: name.clone(),
                 }))
             }
-            _ => return Err(crate::errors::unreachable()),
+            _ => return Err(Unreachable::error().into()),
```

Comment:
I'm seeing we have a is_guard_function not as_guard_function and an unreachable section. I think we should revisit this and either convert it so we don't have an unreachable branch, or add a comment explaining why we can't.